src/DotNetty.Transport/Channels/DefaultChannelId.cs

```diff
@@ -0,0 +1,419 @@
+            byte[] notFound = { byte.MaxValue };
+            byte[] bestMacAddr = notFound;
+            IPAddress bestIpAddr = IPAddress.Loopback;
+            var ifaces = new SortedDictionary<NetworkInterface, IPAddress>();
+            try
+            {
+                foreach (NetworkInterface iface in NetworkInterface.GetAllNetworkInterfaces())
+                {
+                    var addrs = iface.GetIPProperties().UnicastAddresses;
+                    var addr = addrs.FirstOrDefault(a => !IPAddress.IsLoopback(a.Address));
+                    if (addr != null)
+                    {
+                        ifaces.Add(iface, addr.Address);
+                    }
+                }
+            }
+                //catch (SocketException e)
+                //{
+                //    logger.warn("Failed to retrieve the list of available network interfaces", e);
+                //}
+            catch
+            {
+                // ignored
+            }
+
+            foreach (var entry in ifaces)
+            {
+                NetworkInterface iface = entry.Key;
+                IPAddress addr = entry.Value;
+                //todo check if the iface is virtual(there is no equivvalent method in .Net like in java)
+                byte[] macAddr = iface.GetPhysicalAddress().GetAddressBytes();
+                bool replace = false;
+                int res = CompareAddresses(bestMacAddr, macAddr);
+                if (res < 0)
+                {
+                    replace = true;
+                }
+                else if (res == 0)
+                {
+                    res = CompareAddresses(bestIpAddr, addr);
+                    if (res < 0)
+                    {
+                        replace = true;
+                    }
+                    else if (res == 0)
+                    {
+                        if (bestMacAddr.Length < macAddr.Length)
+                        {
+                            replace = true;
+                        }
+                    }
+                }
+
+                if (replace)
+                {
+                    bestMacAddr = macAddr;
+                    bestIpAddr = addr;
+                }
+            }
+
+            if (bestMacAddr == notFound)
+            {
+                bestMacAddr = new byte[MachineIdLen];
+                ThreadLocalRandom.Value.NextBytes(bestMacAddr);
+            }
+
+            switch (bestMacAddr.Length)
+            {
+                case 6: // EUI-48 - convert to EUI-64
+                    byte[] newAddr = new byte[MachineIdLen];
+                    Array.Copy(bestMacAddr, 0, newAddr, 0, 3);
+                    newAddr[3] = 0xFF;
+                    newAddr[4] = 0xFE;
+                    Array.Copy(bestMacAddr, 3, newAddr, 5, 3);
+                    bestMacAddr = newAddr;
+                    break;
+                default: // Unknown
+                    bestMacAddr = bestMacAddr.Take(MachineIdLen).ToArray();
```

Comment:
no action needed, just saying that LINQ would not be acceptable if it was for a code in hot path